hw9: advanced oop/hw9.py

```diff
@@ -0,0 +1,212 @@
+
+# to avoid annoying nested list let's flatten them
+def flatten_list(nested_list):
+    flattened_list = []
+    for element in nested_list:
+        if isinstance(element, list):
+            flattened_list += flatten_list(element)
+        else:
+            flattened_list.append(element)
+    return flattened_list
+
+
+class Args:
+    def __init__(self, *args, **kwargs):
+        self.args = list(args)
+        self.args = flatten_list(self.args)
+        self.kwargs = list(kwargs.values())
+
+        self.all_argies = self.args + self.kwargs
+
+    def __rlshift__(self, other):
+        return other(*self.all_argies)
+
+
+# task 3: a new way to pass arguments to functions
+
+class StrangeFloat(float):
+    def __getattr__(self, attr):
+        if '__' + attr.split('_')[0][:3] + '__' in dir(float):
+            fnc = getattr(float, '__' + attr.split('_')[0][:3] + '__')
+            return fnc(self, float(attr.split('_')[1]))
+        else:
+            dict_attr = {'subtract': self - float(attr.split('_')[1]),
+                         'multiply': self * float(attr.split('_')[1]),
+                         'divide': self / float(attr.split('_')[1])}
+            return StrangeFloat(dict_attr[attr.split('_')[0]])
+
+        def __repr__(self):
+            return float(self)
+
+
+# task 4: rewrite changing to dunders
+
+# overall, 3 (three) dunders
+import numpy as np
+
+matrix2 = []
+for idx in range(0, 100, 10):
+    matrix2.__iadd__([list(range(idx, idx + 10))])
+
+selected_columns_indices = list(filter(lambda x: x in range(1, 5, 2), range(matrix.__len__())))
+selected_columns = map(lambda x: [x.__getitem__(col) for col in selected_columns_indices], matrix)
+
+arr = np.array(list(selected_columns))
+
+mask = arr[:, 1] % 3 == 0
+new_arr = arr[mask]
+
+product = new_arr @ new_arr.T
+
+if (product[0] < 1000).all() and (product[2] > 1000).any():
+    print(product.mean())
+
+
+# task 5:
+
+from abc import ABC, abstractmethod
+
+
+class BiologicalSequence(ABC):
+    def __init__(self, sequence):
+        self.sequence = sequence.upper()
+
+    def __len__(self):
+        return len(self.sequence)
+
+    @abstractmethod
+    def __getitem__(self, slc):
+        pass
+
+    @abstractmethod
+    def check_alphabet(self):
+        pass
+
+    def __repr__(self):
+        return str(self.sequence)
+
+
+class NucleicAcidSequence(BiologicalSequence):
+    def check_alphabet(self):
+        check_na = (['U'], ['T'])[int(self.sequence_type == "DNA")]
+        check_set = set(['A', 'G', 'C'] + check_na)
+        len_foreign = len(set(self.sequence) - check_set)
+        assert len_foreign == 0, 'wrong alphabet!!!'
+
+    def complement(self):
+        if self.sequence_type == "DNA":
+            dict_trans = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C'}
+        elif self.sequence_type == 'RNA':
+            dict_trans = {'A': 'U', 'U': 'A', 'C': 'G', 'G': 'C'}
+        return ''.join([dict_trans[nucleotid] for nucleotid in self.sequence])
```

Comment:
1. Гипотетическая ситуация. У нас есть 50 разных видов нуклеиновых кислот. Писать 50 if-elif это совсем не круто. Гораздо удобнее для каждого из комплементируемых объектов сделать атрибут `dict_trans` и использовать его здесь, а отсюда это лучше убрать. В таком случае получится что-то вот такое, это гораздо гибче
2. Когда мы вызываем `complement` от ДНК, мы ожидаем получить ДНК, когда мы вызываем `complement` от РНК, мы ожидаем получить РНК

 ```suggestion
        complement_string = ''.join([self.dict_trans[nucleotid] for nucleotid in self.sequence])
        return type(self)(complement_string)
```